clients/extension/scripts/popup.js

```diff
@@ -150,160 +59,132 @@ function createOptionItem(text, id, isChecked) {
-    lanternCheck.checked = !settings.useDevTools;
+function persistSettings() {
+  const optionsEl = find('.section--options');
+  // Save settings when options page is closed.
+  const checkboxes = /** @type {NodeListOf<HTMLInputElement>} */
+    (optionsEl.querySelectorAll('.options__categories input:checked'));
+  const selectedCategories = Array.from(checkboxes).map(input => input.value);
+  const device = /** @type {HTMLInputElement} */ (find('input[name="device"]:checked')).value;
+
+  ExtensionController.saveSettings({
+    selectedCategories,
+    device,
   });
+}
 
-  // bind throttling control button
-  const lanternCheckbox = /** @type {HTMLInputElement} */ (find('#lantern-checkbox'));
-  lanternCheckbox.addEventListener('change', async () => {
-    const settings = await background.loadSettings();
-    settings.useDevTools = !lanternCheckbox.checked;
-    background.saveSettings(settings);
-  });
+function getSiteUrl() {
+  return new Promise((resolve, reject) => {
+    chrome.tabs.query({active: true, lastFocusedWindow: true}, function(tabs) {
+      if (tabs.length === 0 || !tabs[0].url) {
+        return;
+      }
 
-  // bind Generate Report button
-  const generateReportButton = find('#generate-report');
-  generateReportButton.addEventListener('click', () => {
-    background.loadSettings().then(settings => {
-      onGenerateReportButtonClick(background, settings);
+      const url = new URL(tabs[0].url);
+      if (url.hostname === 'localhost') {
+        reject(new Error('Use DevTools to audit pages on localhost.'));
+      } else if (/(chrome|chrome-extension):/.test(url.protocol)) {
+        reject(new Error(`Cannot audit ${url.protocol}// pages.`));
+      } else {
+        resolve(url);
+      }
     });
   });
+}
 
-  // bind View Options button
-  const generateOptionsEl = find('#configure-options');
-  const optionsEl = find('.options');
-  generateOptionsEl.addEventListener('click', () => {
-    optionsEl.classList.add(subpageVisibleClass);
-  });
+/**
+ * Initializes the popup's state and UI elements.
+ */
+async function initPopup() {
+  logVersion();
+  fillDevToolsShortcut();
 
-  // bind Save Options button
-  const okButton = find('#ok');
-  okButton.addEventListener('click', () => {
-    // Save settings when options page is closed.
-    const checkboxes = /** @type {NodeListOf<HTMLInputElement>} */
-      (optionsEl.querySelectorAll(':checked'));
-    const selectedCategories = Array.from(checkboxes).map(input => input.value);
+  const bodyEl = find('body');
+  const optionsEl = find('.button--configure');
+  const generateReportButton = /** @type {HTMLButtonElement} */ (find('.button--generate'));
+  const psiDisclaimerEl = find('.psi-disclaimer');
+  const errorMessageEl = find('.errormsg');
+  const optionsFormEl = find('.options__form');
 
-    background.saveSettings({
-      useDevTools: !lanternCheckbox.checked,
-      selectedCategories,
+  try {
+    const siteUrl = await getSiteUrl();
```

Comment:
it's difficult to follow here since it's relying on exceptions for the control flow and it's hard to tell from outside `getSiteUrl` what the two paths are.

One option would be to split this into two functions, or maybe have a 
```
try {
  const siteUrl = await getSiteUrl();
} catch (err) {
  handleUnauditableUrl(err);
}

// Generate checkboxes from saved settings.
// the rest of the function
```
or whatever

Reply:
or split into two functions, yeah, or not use exceptions, but have `getSiteUrl` return the url and have this function deal with it, or some combo of those ideas, or... :)